src/service/resize-observer.js

```diff
@@ -0,0 +1,167 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {pushIfNotExist, removeItem} from '../utils/array';
+import {
+  registerServiceBuilder,
+  registerServiceBuilderInEmbedWin,
+} from '../service';
+import {rethrowAsync} from '../log';
+
+/** @typedef {function(../layout-rect.LayoutSizeDef)} */
+let ObserverCallbackDef;
+
+/**
+ * Pools resize observers. Eventually the hope is that the call sites can
+ * instantiate their own resize observers. But currently there are some
+ * performance issues and pooling is desired. For more info, see
+ * https://bugs.chromium.org/p/chromium/issues/detail?id=1157544
+ *
+ * @implements {../service.Disposable}
+ */
+export class ResizeObserverImpl {
+  /**
+   * @param {!Window} win
+   */
+  constructor(win) {
+    if (!RESIZE_OBSERVER) {
+      return;
+    }
+
+    /** @private @const {!ResizeObserver} */
+    this.observer_ = new win.ResizeObserver((e) => this.processEntries_(e));
+
+    /** @private {?WeakMap<!Element, !../layout-rect.LayoutSizeDef>} */
+    this.lastSizes_ = new WeakMap();
+
+    /** @private {?WeakMap<!Element, !ObserverCallbackDef>} */
+    this.callbacks_ = new WeakMap();
+
+    /** @private @const */
+    this.microtask_ = Promise.resolve();
+  }
+
+  /** @override */
+  dispose() {
+    if (!RESIZE_OBSERVER) {
+      return;
+    }
+    this.callbacks_ = null;
+    this.lastSizes_ = null;
+    this.observer_.dispose();
+  }
+
+  /**
+   * @param {!Element} element
+   * @param {!ObserverCallbackDef} callback
+   */
+  observe(element, callback) {
+    if (!RESIZE_OBSERVER) {
+      return;
+    }
+    let callbacks = this.callbacks_.get(element);
+    if (!callbacks) {
+      callbacks = [];
+      this.callbacks_.set(element, callbacks);
+      this.observer_.observe(element);
+    }
+    if (!pushIfNotExist(callbacks, callback)) {
+      return;
+    }
+    const lastSize = this.lastSizes_.get(element);
```

Comment:
Can we ensure the `lastSize` logic only occurs in the case of an element with a connected concurrent observer.
I want to protect from a very stale value being used and then quickly corrected after observation. 

E.g.:
1. Observe el1. After some time, unobserve el1
2. While not observed, size changes happen to el1
3. Observe el1 again. The stale `lastSize` is used initially.

Reply:
I clean out the `lastSize` after the `unobserve` call, thus there shouldn't be stale values here.